File: data/Experiment1/sparsel0.py
import numpy as np
from itertools import count

class sparse_l0:
    def __init__(self, A, b, penalty):
        self.A = A
        self.b = b
        self.penalty = penalty  # lambda
        self.L = np.max(np.abs(np.dot(A.T, A)))  # Lipschitz constant
        self.sample_size = A.shape[1]
        # initialization
        x = np.ones(self.sample_size)
        self.lengthx = len(x)
        self.x = x / np.sum(x)
        self.objective_list = [self.func_grad(variable=self.x)[0]]

    def ABPG_gain(self, tol=1e-6):
        rho = 1.2
        Gmin = 1e-2
        G = -1
        theta = 1
        gamma = 2
        z = self.x
        objective, a = self.func_grad(variable=self.x)
        for k in range(2000):
            Mk = max(G / rho, Gmin)
            theta_1 = theta
            for t in range(10000):
                G1 = Mk * rho ** t
                if k > 0:
                    theta = self.solve_theta(theta_1, gamma, G1 / G)
                y = (1 - theta) * self.x + theta * z
                fy, grad_fy = self.func_grad(variable=y)
                z1 = self.prox_map(z, grad_fy, G1 * theta ** (gamma - 1) * self.L)
                x1 = (1 - theta) * self.x + theta * z1
                fx, _ = self.func_grad(variable=x1)
                if fx <= fy + np.dot(grad_fy, (x1 - y)) + G1 * theta ** gamma * self.L * np.sum(z1 * np.log(z1 / z)):
                    break
            z = z1
            self.x = x1
            G = G1
            new_objective, a = self.func_grad(variable=x1)
            self.objective_list.append(new_objective)

            if k % 50 == 0:
                print(f"Iteration: {k}, objective value = {new_objective}, gradient norm = {np.linalg.norm(a)}.")
            if np.abs(new_objective - objective) < tol:
                print('The initialization step finishes!')
                print(f"The current iteration is {k}, the objective value is {new_objective}.")
                break
            objective = new_objective

    def step(self):
        print("Optimization begins! The initialization step starts. ")
        # Initialization step
        self.ABPG_gain()

        print('The L0BPG step starts.')
        # L0BPG step
        idx = 1
        objective = self.objective_value()
        for k in count():
            # BPG step
            y_new, _ = self.iteration()

            # sorting step
            idx, sample_size = self.search_dk(y_new, k, idx)
            remove_num = sample_size - idx
            if remove_num != 0:
                if k != 0:
                    # recover the vector with zero values
                    ynnew = np.zeros(self.lengthx)
                    # for i in range(len(ind)):
                    #     index = ind[i]
                    #     ynnew[index] = y_new[i]
                    ynnew[ind] = y_new
                    y_new = ynnew
                indx = np.argsort(y_new)
                ind = indx[self.lengthx - idx:]

                # removing step
                xm = y_new[ind]
                self.x = xm / np.sum(xm)
                self.A = self.A[:, ind]
            else:
                self.x = y_new
            new_objective = self.objective_value()
            self.objective_list.append(new_objective)
            if np.abs(objective - new_objective) < 1e-6:
                # recover final solution
                xnnew = np.zeros(self.lengthx)
                # for i in range(len(ind)):
                #     index = ind[i]
                #     xnnew[index] = self.x[i]
                xnnew[ind] = self.x
                self.x = xnnew
                print(f"Finish! The current iteration is {k}, the objective value is {new_objective} "
                      f"with nonzero number {np.sum(self.x>0)}.")
                break
            objective = new_objective
            if k % 10 == 0:
                print(f"Iteration: {k}, objective value = {new_objective}, nonzero number = {np.sum(self.x>0)}.")

    def func_grad(self, variable):
        '''
        calculate the objective value and the gradient
        :param variable:
        :return: objective value and gradient
        '''
        residual = np.dot(self.A, variable) - self.b
        grad = np.dot(self.A.T, residual)
        obj = 0.5 * np.linalg.norm(residual) ** 2
        return obj, grad

    def prox_map(self, variable, a, L):
        '''
        explicit expression for x
        '''
        x_without_n = variable * np.exp(-a / L)
        x = x_without_n / np.sum(x_without_n)
        return x

    def objective_value(self):
        obj = 0.5 * np.linalg.norm(np.dot(self.A, self.x) - self.b) ** 2 + self.penalty * np.sum(self.x != 0)
        return obj

    def solve_theta(self, theta, gamma, gainratio=1):
        # (1 - theta_k1) / theta_k1 ^ gamma = gainratio * 1 / theta_k ^ gamma
        ckg = theta ** gamma / gainratio
        cta = theta
        eps = 1e-6 * theta
        phi = cta ** gamma - ckg * (1 - cta)
        while abs(phi) > eps:
            drv = gamma * cta ** (gamma - 1) + ckg
            cta = cta - phi / drv
            phi = cta ** gamma - ckg * (1 - cta)
        return cta

    def iteration(self, num=1):
        y_new = self.x
        new_objective, a = self.func_grad(variable=self.x)
        for _ in range(num):
            y_new = self.prox_map(self.x, a, self.L)
            new_objective, a = self.func_grad(variable=y_new)
            # objective = new_objective
        return y_new, new_objective

    def search_dk(self, x_new, k, dk_last):
        '''
        search d_k in the sorting step
        '''
        a_sort = np.flip(np.sort(x_new[x_new > 0]))
        sample_size = len(a_sort)
        threshold = np.exp(self.penalty / self.L) - 1
        if k == 0:
            denominator = 0
            for m in range(1, sample_size):
                denominator += a_sort[m - 1]
                fraction = a_sort[m] / denominator
                if threshold > fraction:
                    return m, sample_size
        else:
            denominator = np.sum(a_sort)
            for m in range(dk_last, 0, -1):
                denominator -= a_sort[m - 1]
                fraction = a_sort[m - 1] / denominator
                if threshold <= fraction:
                    return m, sample_size

File: sparse_l0.m
function [out] = sparse_l0(A,lambda, b,num, tol,hist)
tstart=tic;
L = max(max(abs(A'*A)));

[~,sample_size] = size(A);
x=ones(sample_size,1);
lengthx = length(x);

x = x/sum(x);
out.history = x;

A1 = A;
[o,x,~,new_objective,~] = ABPG_gain(L,A,x,b,tol, hist);
out.F = o.F;


k = 1;
idx = 1;
objective = objective_value(A,x,b,lambda);
while 1
    % BGP step
    [o,y_new,~] = iteration(L,A1,x,b,lambda,num);

    % sorting step
    [idx,sample_size] = search_dk(y_new,L,lambda,k,idx);
    remove_num = sample_size - idx;

    % removing step
    if remove_num~=0
        % recover x_new
        if k ~= 1
            ynnew = zeros(lengthx,1);
            for i = 1:length(ind)
                index = ind(i);
                ynnew(index) = y_new(i);
            end
            y_new = ynnew;
        end

        [~,indx] = sort(y_new);

        ind = indx(lengthx-idx+1:end);
        
        xm = y_new(ind);
    
        x_new = xm/sum(xm);
    
        % remove the zero entries, accelerate the algorithm
        A1 = A(:,ind);

    else
        x_new = y_new;
    end

    new_objective = objective_value(A1,x_new,b,lambda);

    if abs(objective- new_objective)<tol
        % recover x_new
        xnnew = zeros(lengthx,1);
        for i = 1:length(ind)
            index = ind(i);
            xnnew(index) = x_new(i);
        end
        if hist
            fprintf('total iter\t%d\n',k)
        end
        break
    end
    k = k + 1;
    x = x_new;
    objective = new_objective;
    
    if hist
        if mod(k,50) == 0
            fprintf('iter\t%d:\t F(x) = %e\t nonzero = %d\n',[k,new_objective,sum(x_new~=0)])
        end
    end
end

if hist
    fprintf('-------------Ends--------------\n')
end
time=toc(tstart);
out.time = time;
out.num = k;

% we have attain sparse result
out.x = xnnew;
out.objective = new_objective;
if hist
    fprintf('We have the final result the nonzero number is %d\n' ...
        ,sum(xnnew~=0))     
end
% fprintf('--------------Ends--------------\n')

end

function [idx,sample_size] = search_dk(x_new,L,lambda,k,dk_last)
    a_sort = flip(sort(x_new(x_new > 0)));
    sample_size = length(a_sort);
    threshold = exp(lambda/L)-1;
    if k == 1
        denominator = 0;
        for m = 1:(sample_size-1)
            denominator = denominator + a_sort(m);
            fraction = a_sort(m+1)/denominator;
            if threshold > fraction
                idx = m;
                break;
            end
        end
    else
        denominator = sum(a_sort);
        for m = (dk_last):-1:1
            denominator = denominator - a_sort(m);
            fraction = a_sort(m)/denominator;
            if threshold <= fraction
                idx = m;
                break;
            end
        end
    end
end



function [out,x,z1,new_objective,ind] = ABPG_gain(L,A,x,b,tol,hist)
rho = 1.2;
% out.x = [];

Gmin = 1e-2;
G = -1; theta = 1; gamma = 2;
z = x;
out.F = [];
% out.theta = [];
[objective,~] = func_grad(A,x,b);
A1 = A;
ind = [];

for k = 1:2000
    Mk = max(G/rho,Gmin);
    theta_1 = theta;
    out.F{k} = objective;
    for t = 1:10000
        G1 = Mk*rho^(t-1);
        if k>1
            theta = solve_theta(theta_1,gamma,G1/G);
        end
        y = (1-theta) * x + theta * z;
        [fy,grad_fy] = func_grad(A1,y,b);
        z1 = prox_map(z,grad_fy,G1*theta^(gamma-1)*L);
        x1 = (1-theta)*x + theta*z1;
        [fx,~] = func_grad(A1,x1,b);
        if fx <= fy + sum(grad_fy.*(x1 - y)) + G1*theta^(gamma)*L*...
                sum(z1.*log(z1./z))
            break
        end
    end

    

    % out.theta{k} = theta;
    z = z1;
    x = x1;
    G = G1;
    [new_objective,a] = func_grad(A1,x1,b);
    out.obj(k) = new_objective;
    if hist
        if mod(k,100) == 0
            fprintf('iter\t%d:\t F(x) = %e \t gradient = %e\n',[k,new_objective,norm(a)])
    
        end
    end

    if abs(new_objective - objective) < tol % abs(new_objective-objective)<tol 
        if hist
            fprintf('subproblem solved with iteration %d-------------Ends--------------\n',k)
        end
        break
    end


    objective = new_objective;
    % out.x{k} = x;
    % out.z{k} = z;
end
end


function [out,x,new_objective] = iteration(L,A,x,b,lambda,num)
[objective,a] = func_grad(A,x,b);
for i = 1:num
    out.F{i} = objective + lambda*sum(x~=0);
    % under the fixed lambda, updated the x
    x = prox_map(x,a,L);
    [new_objective,a] = func_grad(A,x,b);
    objective = new_objective;
end
end



function x = prox_map(x,a,L)
x_without_n = x.*exp(-a/L); 
x = x_without_n/sum(x_without_n);
end

function [obj,grad] = func_grad(A,x,b)
residual = A*x-b;
grad = A'*residual;
obj = 1/2*norm(residual,2)^2;
end

function [obj] = objective_value(A,x,b,lambda)
obj = 1/2*norm(A*x-b,2)^2 + lambda*sum(x~=0);
end

function cta = solve_theta(theta,gamma,gainratio)
% (1-theta_k1)/theta_k1^gamma = gainratio * 1/theta_k^gamma
ckg = theta^gamma / gainratio;
cta = theta;
eps = 1e-6 * theta;
phi = cta^gamma - ckg*(1-cta);
while abs(phi) > eps
    drv = gamma * cta^(gamma-1) + ckg;
    cta = cta - phi / drv;
    phi = cta^gamma - ckg*(1-cta);
end

end
